File: algorithms/06_javascript/data_structures/avl_tree.js
// Self-balancing AVL tree.
class Node {
  constructor(key) {
    this.key = key;
    this.height = 1;
    this.left = null;
    this.right = null;
  }
}

const height = (node) => (node ? node.height : 0);
const update = (node) => {
  node.height = 1 + Math.max(height(node.left), height(node.right));
};

function rotateRight(y) {
  const x = y.left;
  y.left = x.right;
  x.right = y;
  update(y);
  update(x);
  return x;
}

function rotateLeft(x) {
  const y = x.right;
  x.right = y.left;
  y.left = x;
  update(x);
  update(y);
  return y;
}

function balance(node) {
  update(node);
  const factor = height(node.left) - height(node.right);
  if (factor > 1) {
    if (height(node.left.left) < height(node.left.right)) node.left = rotateLeft(node.left);
    return rotateRight(node);
  }
  if (factor < -1) {
    if (height(node.right.right) < height(node.right.left)) node.right = rotateRight(node.right);
    return rotateLeft(node);
  }
  return node;
}

function insert(node, key) {
  if (!node) return new Node(key);
  if (key < node.key) node.left = insert(node.left, key);
  else if (key > node.key) node.right = insert(node.right, key);
  else return node;
  return balance(node);
}

function inorder(node, out = []) {
  if (!node) return out;
  inorder(node.left, out);
  out.push(node.key);
  inorder(node.right, out);
  return out;
}

let root = null;
for (const key of [10, 20, 30, 40, 50, 25]) root = insert(root, key);
const out = inorder(root);
if (out.some((value, index) => index && out[index - 1] > value)) throw new Error("not sorted");
console.log("avl tree ok");

File: algorithms/11_golang/data_structures/avl_tree.go
// Self-balancing AVL tree.
package main

import "fmt"

type node struct {
	key, height  int
	left, right  *node
}

func height(n *node) int {
	if n == nil {
		return 0
	}
	return n.height
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func update(n *node) { n.height = 1 + max(height(n.left), height(n.right)) }

func rotateRight(y *node) *node {
	x := y.left
	y.left = x.right
	x.right = y
	update(y)
	update(x)
	return x
}

func rotateLeft(x *node) *node {
	y := x.right
	x.right = y.left
	y.left = x
	update(x)
	update(y)
	return y
}

func balance(n *node) *node {
	update(n)
	factor := height(n.left) - height(n.right)
	if factor > 1 {
		if height(n.left.left) < height(n.left.right) {
			n.left = rotateLeft(n.left)
		}
		return rotateRight(n)
	}
	if factor < -1 {
		if height(n.right.right) < height(n.right.left) {
			n.right = rotateRight(n.right)
		}
		return rotateLeft(n)
	}
	return n
}

func insert(n *node, key int) *node {
	if n == nil {
		return &node{key: key, height: 1}
	}
	if key < n.key {
		n.left = insert(n.left, key)
	} else if key > n.key {
		n.right = insert(n.right, key)
	} else {
		return n
	}
	return balance(n)
}

func inorder(n *node, out *[]int) {
	if n == nil {
		return
	}
	inorder(n.left, out)
	*out = append(*out, n.key)
	inorder(n.right, out)
}

func main() {
	var root *node
	for _, key := range []int{10, 20, 30, 40, 50, 25} {
		root = insert(root, key)
	}
	var out []int
	inorder(root, &out)
	for i := 1; i < len(out); i++ {
		if out[i-1] > out[i] {
			panic("not sorted")
		}
	}
	fmt.Println("avl tree ok")
}
